### ISSUE TITLE
Consider Memory Efficient Attention as an "alternative" to Flash Attention for AMD users.

### ISSUE THREAD

**Issue by xzuyn** [axolotl-ai-cloud/axolotl#1519]
### ⚠️ Please check that this feature request hasn't been suggested before.

- [X] I searched previous [Ideas in Discussions](https://github.com/OpenAccess-AI-Collective/axolotl/discussions/categories/ideas) didn't find any similar feature requests.
- [X] I searched previous [Issues](https://github.com/OpenAccess-AI-Collective/axolotl/labels/enhancement) didn't find any similar feature requests.

### 🔖 Feature description

I would like to see [lucidrain's PyTorch implementation of Memory Efficient Attention](https://github.com/lucidrains/memory-efficient-attention-pytorch) added/tested within Axolotl. This has the possibility of reducing memory usage by ~2.1x, with the trade-off of slowing down the training speed by ~2.4x. Those numbers come from stable diffusion training within kohya_ss, and I've included more details in the **Additional Context** section.

kohya_ss uses lucidrain's code within these files:
- [attention_processors.py](https://github.com/kohya-ss/sd-scripts/blob/main/library/attention_processors.py)
- [original_unet.py](https://github.com/kohya-ss/sd-scripts/blob/main/library/original_unet.py)
- [sdxl_original_unet.py](https://github.com/kohya-ss/sd-scripts/blob/main/library/sdxl_original_unet.py)

### ✔️ Solution

I would like someone to test lucidrain's PyTorch implementation of Memory Efficient Attention (or any other PyTorch implementation of Memory Efficient Attention or Flash Attention 1/2 which doesn't require specific hardware to work) within Axolotl to determine whether the memory improvements seen in kohya_ss also occur in Axolotl or if they are merely a fluke/issue with kohya_ss itself (or an issue with my benchmarking, which is entirely possible).

If it's not a fluke, I would love to see it added here to assist users who encounter memory issues even with relatively low training settings compared to what can be achieved with Flash Attention.

Despite the much slower speeds, I personally find this trade-off worthwhile as it would enable the use of much larger models or sequence lengths than were previously feasible on AMD cards if the user was willing to wait.

### 📝 Additional Context

During my testing with it in kohya_ss, it greatly reduced memory usage, although it did increase training time which is why I titled this with "alternative" in quotes. I have not done any tests with Axolotl or any LLMs, only with kohya_ss.

I tested with it enabled and disabled, both using SDPA so they should both be defaulting to the best available kernel, yet the Memory Efficient Attention still lowers the memory usage by a lot. Which in my version of PyTorch; `torch=2.3.0.dev20240221+rocm6.0`, is `enable_math_sdp()` as far as I know.

These two screenshots are using the same training config, with the only difference being Memory Efficient Attention being enabled or disabled. Both are running on my 7900 XTX.
```
LoRA Rank: 128
Image size: 768x768 (no buckets)
Batch Size: 1
Gradient Checkpointing: True
CrossAttention: SDPA
Mixed Precision: BF16
Optimizer: PagedAdamW
Cache Latents: True
Cache Latents to Disk: True

Memory Efficient Attention: Disabled

Speed: ~1.7it/s
Memory Usage: ~11.3GB
```
![Screenshot from 2024-04-14 20-47-27](https://github.com/OpenAccess-AI-Collective/axolotl/assets/16216325/5ac0f152-93b2-490c-9c9d-2d0aa2f8b14a)

```
LoRA Rank: 128
Image size: 768x768 (no buckets)
Batch Size: 1
Gradient Checkpointing: True
CrossAttention: SDPA
Mixed Precision: BF16
Optimizer: PagedAdamW
Cache Latents: True
Cache Latents to Disk: True

Memory Efficient Attention: Enabled

Speed: ~1.4s/it (~2.38x slower)
Memory Usage: ~5.4GB (~2.09x less)
```
![Screenshot from 2024-04-14 20-46-31](https://github.com/OpenAccess-AI-Collective/axolotl/assets/16216325/c22639b3-b75f-4bb4-bf22-0e6793da8033)

### Acknowledgements

- [X] My issue title is concise, descriptive, and in title casing.
- [X] I have searched the existing issues to make sure this feature has not been requested yet.
- [X] I have provided enough information for the maintainers to understand and evaluate this request.

**Comment by xzuyn** [comment]
Not needed anymore as PyTorch added `TORCH_ROCM_AOTRITON_ENABLE_EXPERIMENTAL` a while back, and there is a *mostly* working ROCm fork of Flash Attention.
